cloudinit/sources/DataSourceVMware.py

```diff
@@ -776,91 +774,52 @@ def get_default_ip_addrs():
     addresses associated with the device used by the default route for a given
     address.
     """
-    # TODO(promote and use netifaces in cloudinit.net* modules)
-    gateways = netifaces.gateways()
-    if "default" not in gateways:
-        return None, None
-
-    default_gw = gateways["default"]
-    if (
-        netifaces.AF_INET not in default_gw
-        and netifaces.AF_INET6 not in default_gw
-    ):
-        return None, None
-
     ipv4 = None
     ipv6 = None
+    routes = netinfo.route_info()
+    for route in routes:
+        if "G" in route["flags"]:
+            if ipv4 and ipv4 != route["gateway"]:
+                LOG.debug(
+                    "multiple ipv4 gateways: %s, %s", ipv4, route["gateway"]
+                )
+            ipv4 = route["gateway"]
+        elif "UG" in route["flags"]:
+            if ipv6 and ipv6 != route["gateway"]:
+                LOG.debug(
+                    "multiple ipv6 gateways: %s, %s", ipv6, route["gateway"]
+                )
+            ipv6 = route["gateway"]
 
-    gw4 = default_gw.get(netifaces.AF_INET)
-    if gw4:
-        _, dev4 = gw4
-        addr4_fams = netifaces.ifaddresses(dev4)
-        if addr4_fams:
-            af_inet4 = addr4_fams.get(netifaces.AF_INET)
-            if af_inet4:
-                if len(af_inet4) > 1:
-                    LOG.debug(
-                        "device %s has more than one ipv4 address: %s",
-                        dev4,
-                        af_inet4,
-                    )
-                elif "addr" in af_inet4[0]:
-                    ipv4 = af_inet4[0]["addr"]
-
-    # Try to get the default IPv6 address by first seeing if there is a default
-    # IPv6 route.
-    gw6 = default_gw.get(netifaces.AF_INET6)
-    if gw6:
-        _, dev6 = gw6
-        addr6_fams = netifaces.ifaddresses(dev6)
-        if addr6_fams:
-            af_inet6 = addr6_fams.get(netifaces.AF_INET6)
-            if af_inet6:
-                if len(af_inet6) > 1:
-                    LOG.debug(
-                        "device %s has more than one ipv6 address: %s",
-                        dev6,
-                        af_inet6,
-                    )
-                elif "addr" in af_inet6[0]:
-                    ipv6 = af_inet6[0]["addr"]
+    # If there is an IPv4 gateway and an IPv6 gateway, return immediately to
+    # avoid extra work
+    if ipv4 and ipv6:
+        return ipv4, ipv4
```

Comment:
This function is to get ipv4 and ipv6 addresses of default interface, so return gateway is not expected. And it should be `return ipv4, ipv6`

Reply:
I think I have misunderstood the goal of this function.  Based on your comments, the correct process for this functions is (one flow but actually IPv4/IPv6 for each step):
1. Identify default gateway
2. Extract interface of default gateway
3. Get IP address of interface
4. Return IP address

I will update the code to reflect this flow unless you have further corrections to add.  Thanks again for your input here!